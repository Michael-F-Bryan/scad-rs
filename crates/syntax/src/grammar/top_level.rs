use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
