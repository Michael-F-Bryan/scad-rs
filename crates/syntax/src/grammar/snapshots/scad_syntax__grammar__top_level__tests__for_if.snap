---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(for(i = [0: 1]) if (true) cube(i);)"
---
Node(
    FOR_STATEMENT@0..34
      FOR_KW@0..3 "for"
      L_PAREN@3..4 "("
      ASSIGNMENTS@4..14
        ASSIGNMENT@4..14
          IDENT@4..5 "i"
          WHITESPACE@5..6 " "
          EQUALS@6..7 "="
          RANGE_EXPR_FROM_TO@7..14
            WHITESPACE@7..8 " "
            L_BRACKET@8..9 "["
            INTEGER@9..10
              INTEGER@9..10 "0"
            COLON@10..11 ":"
            INTEGER@11..13
              WHITESPACE@11..12 " "
              INTEGER@12..13 "1"
            R_BRACKET@13..14 "]"
      R_PAREN@14..15 ")"
      IF_STATEMENT@15..34
        WHITESPACE@15..16 " "
        IF_KW@16..18 "if"
        WHITESPACE@18..19 " "
        L_PAREN@19..20 "("
        TRUE_KW@20..24
          TRUE_KW@20..24 "true"
        R_PAREN@24..25 ")"
        MODULE_INSTANTIATION@25..34
          WHITESPACE@25..26 " "
          IDENT@26..30 "cube"
          L_PAREN@30..31 "("
          ARGUMENTS@31..32
            LOOKUP_EXPR@31..32
              IDENT@31..32 "i"
          R_PAREN@32..33 ")"
          SEMICOLON@33..34
            SEMICOLON@33..34 ";"
    ,
)
